cardinal/world.go

```diff
@@ -0,0 +1,168 @@
+package cardinal
+
+import (
+	"context"
+	"errors"
+	"time"
+
+	"github.com/rs/zerolog/log"
+	"pkg.world.dev/world-engine/cardinal/ecs"
+	"pkg.world.dev/world-engine/cardinal/ecs/inmem"
+	"pkg.world.dev/world-engine/cardinal/ecs/storage"
+	"pkg.world.dev/world-engine/cardinal/ecs/transaction"
+	"pkg.world.dev/world-engine/cardinal/server"
+	"pkg.world.dev/world-engine/cardinal/shard"
+)
+
+type World struct {
+	impl *ecs.World
+}
+
+// System is a function that process the transaction in the given transaction queue.
+// Systems are automatically called during a world tick, and they must be registered
+// with a world using AddSystem or AddSystems.
+type System func(*World, *TransactionQueue) error
+
+type (
+	// EntityID represents a single entity in the World. An EntityID is tied to
+	// one or more components.
+	EntityID = storage.EntityID
+	TxHash   = transaction.TxHash
+)
+
+// NewWorldUsingRedis creates a new World object using Redis as the storage layer.
+func NewWorldUsingRedis(addr, password string, opts ...WorldOption) (*World, error) {
+	log.Log().Msg("Running in normal mode, using external Redis")
+	if addr == "" {
+		return nil, errors.New("redis address is required")
+	}
+	if password == "" {
+		log.Log().Msg("Redis password is not set, make sure to set up redis with password in prod")
+	}
+
+	rs := storage.NewRedisStorage(storage.Options{
+		Addr:     addr,
+		Password: password, // make sure to set this in prod
+		DB:       0,        // use default DB
+	}, "world")
+	worldStorage := storage.NewWorldStorage(&rs)
+	world, err := ecs.NewWorld(worldStorage, opts...)
+	if err != nil {
+		return nil, err
+	}
+
+	return &World{
+		impl: world,
+	}, nil
+}
+
+// NewWorldInMemory creates a World that uses an in-memory redis DB as the storage
+// layer. This is only suitable for local development.
+func NewWorldInMemory(opts ...WorldOption) (*World, error) {
+	return &World{
+		impl: inmem.NewECSWorld(opts...),
+	}, nil
+}
+
+// AddSystem adds the given system to the world object so that the system will be executed
+// at every game tick.
+func (w *World) AddSystem(system System) {
+	w.impl.AddSystem(func(world *ecs.World, queue *ecs.TransactionQueue) error {
+		return system(&World{world}, &TransactionQueue{queue})
+	})
+}
+
+// AddSystems allows for the adding of multiple systems in a single call. See AddSystem for more details.
+func (w *World) AddSystems(systems ...System) {
```

Comment:
I think @technicallyty wanted to rename this to RegisterSystem to be consistent with RegisterComponent and RegisterReadds

Reply:
Renamed to RegisterSystem and RegsiterSystems. Do we need both? How about just RegisterSystems (the plural) that takes in a variable number of systems. RegisterSystems is a little different from the other Register methods because you're allowed to call it multiple times.